####-sparse-observable.md

```diff
@@ -0,0 +1,517 @@
+The particulars of `Parameter` and `Value` and the containing mapping are irrelevant to this RFC.
+
+`SparseObservable` will support some set of mathematical operations.
+At a minimum, the following will be supported:
+
+* addition of two `SparseObservable`s
+* tensor product of two `SparseObservable`s
+* evolution of one `SparseObservable` by a Pauli string ($A' = P A P^\dagger$ for `SparseObservable` $A$ and Pauli $P$)
+* multiplication by complex scalars
+* structural equality of two `SparseObservable`s (structural not mathematical; it's highly inefficient to detect equality if the abstract terms form an over-complete spanning set).
+
+See "Questions" for more on the mathematical-operator support.
+
+Iterator methods for accessing the terms and their coefficients will be supported.
+The exact format of these is to be informed by the intended use cases, and there can be more than one iterator method available.
+
+
+### Representation of measurement bases
+
+> [!NOTE]
+> **Summary**: Qiskit will supply some helper functions on `SparseObservable`, but for MVP 1 of the primitive internals, we can keep using `PauliList` for this bit, and we can optimise it in the future if needed.
+
+While the user input is a n-dimensional group of observable sums, the actual execution is serialised and only cares about measurement-basis transformation needed for any given multi-qubit term.
+See the "Detailed Design" section for more discussion on exactly how this grouping works.
+
+From Python-space, then, we produce an iterable of "preparation programs" (pairs of `(circuit, bindings)`), and for element of that we have an iterable of `SparseObservable` with we want to map to a group of groups of measurement bases.
+The iterable of preparation programs is produced in Python-space by n-dimensional slicing/broadcasting tricks.
+
+A core function of `Estimator` is to group terms that can be measured within the same execution.
+There are many ways to do this, and we do not want to tie the observable to one particular implementation.
+Qiskit will provide a function `SparseObservable.measurement_bases(*observables)` that takes an arbitrary number of `SparseObservable` instances and returns a set of the measurement bases needed to measure all terms.
+A measurement basis is a Pauli string.
+
+Since the number of measurement bases will be (non-strictly) smaller than the total number of terms across all observables, and because only 2 bits of information per qubit is necessary to define the basis, there is not expected to be immediate memory concerns with a representation of this.
+Qiskit already has `PauliList` that can serve this purpose; it _could_ be bit-packed to use 8x less memory, but this can be done as a follow-up optimisation if it becomes a bottleneck.
+From this point, we can continue to use `PauliList.group_qubitwise_commuting`, or any other future grouping function.
```

Comment:
The only part of `SparsePauliOp` that causes serious immediate memory problems is that some things that are efficient to measure are not efficient to represent.  For example, the all-zeros projector state takes linear complexity/space to measure (it's an all-Z basis) but needs $2^n$ terms in `SparsePauliOp` to represent.  Since, for the measurement, you just need to know which basis to take your counts in, the information about "what to measure" (including mitigation) can be stored efficiently by `PauliList`, and you reconstruct the requested observables later from the original `SparseObservable`.

Bit-packing `PauliList` can reduce its memory usage by a factor of about 8, but that's just a scaling factor.  We already know we _must_ be able to do operations that are linear in the number of qubits (or how would we twirl?), and `PauliList` can represent all the measurement bases we realistically care about for error-mitigation purposes for the next while.

All that said, how a primitives implementation actually manages its error mitigation is entirely up to it, and this isn't fixed by any public interface.  So any primitive can use whatever they like in the backend to do this task; the point about `PauliList` is mostly just showing that we don't immediately need any new object.

Reply:
Thanks @jakelishman! I was thinking more on all the design decisions to avoid explicit identities, but I see your point that this is not the biggest concern 🙂